backend/server/src/services/tournamentService.ts

```diff
@@ -67,4 +100,262 @@ export class TournamentService {
+    tournament.matchSchedule.push(newMatch._id);
+    await tournament.save();
+    return await tournament.toObject();
+  }
+
+  // update - auto generate match
+  public async generateTournamentSchedule(
+    tournamentId: string
+  ): Promise<Tournament> {
+    const tournament = await TournamentModel.findById(tournamentId).exec();
+
+    if (tournament === null || tournament === undefined) {
+      throw new NotFoundError({
+        message: "Tournament not found"
+      });
+    }
+
+    if (tournament.players.length !== tournament.maxPlayers) {
+      throw new BadRequestError({
+        message: "Tournament player count has not reached the maximum capacity"
+      });
+    }
+
+    let matches: UnsavedMatch[] = [];
+
+    switch (tournament.tournamentType) {
+      case TournamentType.RoundRobin:
+        if (
+          this.calculateRoundRobinMatches(tournament.maxPlayers) ===
+          tournament.matchSchedule.length
+        ) {
+          throw new BadRequestError({
+            message:
+              "We have already created the match schedule for this tournament."
+          });
+        }
+        matches = this.generateRoundRobinSchedule(tournament.players);
+        break;
+      case TournamentType.Playoff:
+        matches = await this.generatePlayoffSchedule(
+          tournament.players,
+          tournament.matchSchedule
+        );
+        break;
+    }
+
+    // Create Match documents for each match
+    for (const match of matches) {
+      const matchDocument = await MatchModel.create(match);
+      tournament.matchSchedule.push(matchDocument._id);
+    }
+
+    await tournament.save();
+    return await tournament.toObject();
+  }
+
+  // private function
+  private generateRoundRobinSchedule(
+    playerIds: Types.ObjectId[]
+  ): UnsavedMatch[] {
+    const matches: UnsavedMatch[] = [];
+    for (let i = 0; i < playerIds.length; i++) {
+      for (let j = i + 1; j < playerIds.length; j++) {
+        matches.push({
+          players: [
+            { id: playerIds[i], points: [], color: "red" },
+            { id: playerIds[j], points: [], color: "white" }
+          ],
+          type: "group", // decide what this is going to be.
+          admin: null, // this has to be an ID.
+          elapsedTime: 0,
+          timerStartedTimestamp: null
+          // Set other Match fields as necessary
+        });
+      }
+    }
+    return matches;
+  }
+
+  // private function
+  private async generatePlayoffSchedule(
+    playerIds: Types.ObjectId[],
+    previousMatches: Types.ObjectId[]
+  ): Promise<UnsavedMatch[]> {
+    const matches: UnsavedMatch[] = [];
+    let currentRoundPlayers = [];
+    // check if there's any previous matches.
+    if (previousMatches.length === 0) {
+      // work with original list
+      currentRoundPlayers = [...playerIds];
+    } else {
+      const matchPromises = previousMatches.map(
+        async (matchId) => await MatchModel.findById(matchId).exec()
+      );
```

Comment:
Did you try populating the Matches when you query the initial Tournament? You could then avoid this part completely and just pass the Match collection here